src/peer/chunk/write/state.rs

```diff
@@ -0,0 +1,71 @@
+use std::io;
+
+use serde::{Deserialize, Serialize};
+
+use crypto::{crypto_box::PrecomputedKey, nonce::Nonce, CryptoError};
+use tezos_messages::p2p::binary_message::{BinaryChunk, BinaryChunkError};
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum PeerConnectionChunkWriteState {
+    Init,
+    Pending,
+    Ready,
+    Error(PeerChunkWriteError),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct WriteCrypto {
+    pub precomputed_key: PrecomputedKey,
+    pub local_nonce: Nonce,
+}
+
+impl WriteCrypto {
+    pub fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, CryptoError> {
+        self.precomputed_key.encrypt(data, &self.local_nonce)
+    }
+
+    pub fn increment_nonce(&mut self) {
+        self.local_nonce = self.local_nonce.increment()
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum PeerChunkWriteError {
+    IO(String),
```

Comment:
For now we could keep the `String` errors, but it will need to change later.

Not only it's inefficient, but makes error handling harder.